File: pulsar-third/pulsar-chrome-devtools/src/main/java/com/github/kklisura/cdt/protocol/commands/ApplicationCache.java
package com.github.kklisura.cdt.protocol.commands;

/*-
 * #%L
 * cdt-java-client
 * %%
 * Copyright (C) 2018 - 2019 Kenan Klisura
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.kklisura.cdt.protocol.events.applicationcache.ApplicationCacheStatusUpdated;
import com.github.kklisura.cdt.protocol.events.applicationcache.NetworkStateUpdated;
import com.github.kklisura.cdt.protocol.support.annotations.EventName;
import com.github.kklisura.cdt.protocol.support.annotations.Experimental;
import com.github.kklisura.cdt.protocol.support.annotations.ParamName;
import com.github.kklisura.cdt.protocol.support.annotations.ReturnTypeParameter;
import com.github.kklisura.cdt.protocol.support.annotations.Returns;
import com.github.kklisura.cdt.protocol.support.types.EventHandler;
import com.github.kklisura.cdt.protocol.support.types.EventListener;
import com.github.kklisura.cdt.protocol.types.applicationcache.FrameWithManifest;
import java.util.List;

@Experimental
public interface ApplicationCache {

  /** Enables application cache domain notifications. */
  void enable();

  /**
   * Returns relevant application cache data for the document in given frame.
   *
   * @param frameId Identifier of the frame containing document whose application cache is
   *     retrieved.
   */
  @Returns("applicationCache")
  com.github.kklisura.cdt.protocol.types.applicationcache.ApplicationCache
      getApplicationCacheForFrame(@ParamName("frameId") String frameId);

  /**
   * Returns array of frame identifiers with manifest urls for each frame containing a document
   * associated with some application cache.
   */
  @Returns("frameIds")
  @ReturnTypeParameter(FrameWithManifest.class)
  List<FrameWithManifest> getFramesWithManifests();

  /**
   * Returns manifest URL for document in the given frame.
   *
   * @param frameId Identifier of the frame containing document whose manifest is retrieved.
   */
  @Returns("manifestURL")
  String getManifestForFrame(@ParamName("frameId") String frameId);

  @EventName("applicationCacheStatusUpdated")
  EventListener onApplicationCacheStatusUpdated(
      EventHandler<ApplicationCacheStatusUpdated> eventListener);

  @EventName("networkStateUpdated")
  EventListener onNetworkStateUpdated(EventHandler<NetworkStateUpdated> eventListener);
}

File: pulsar-core/pulsar-tools/pulsar-browser/src/main/kotlin/ai/platon/cdt/kt/protocol/commands/ApplicationCache.kt
@file:Suppress("unused")
package ai.platon.cdt.kt.protocol.commands

import ai.platon.cdt.kt.protocol.events.applicationcache.ApplicationCacheStatusUpdated
import ai.platon.cdt.kt.protocol.events.applicationcache.NetworkStateUpdated
import ai.platon.cdt.kt.protocol.support.annotations.EventName
import ai.platon.cdt.kt.protocol.support.annotations.Experimental
import ai.platon.cdt.kt.protocol.support.annotations.ParamName
import ai.platon.cdt.kt.protocol.support.annotations.ReturnTypeParameter
import ai.platon.cdt.kt.protocol.support.annotations.Returns
import ai.platon.cdt.kt.protocol.support.types.EventHandler
import ai.platon.cdt.kt.protocol.support.types.EventListener
import ai.platon.cdt.kt.protocol.types.applicationcache.FrameWithManifest
import kotlin.String
import kotlin.Unit
import kotlin.collections.List

@Experimental
interface ApplicationCache {
  /**
   * Enables application cache domain notifications.
   */
  suspend fun enable()

  /**
   * Returns relevant application cache data for the document in given frame.
   * @param frameId Identifier of the frame containing document whose application cache is retrieved.
   */
  @Returns("applicationCache")
  suspend fun getApplicationCacheForFrame(@ParamName("frameId") frameId: String): ai.platon.cdt.kt.protocol.types.applicationcache.ApplicationCache

  /**
   * Returns array of frame identifiers with manifest urls for each frame containing a document
   * associated with some application cache.
   */
  @Returns("frameIds")
  @ReturnTypeParameter(FrameWithManifest::class)
  suspend fun getFramesWithManifests(): List<FrameWithManifest>

  /**
   * Returns manifest URL for document in the given frame.
   * @param frameId Identifier of the frame containing document whose manifest is retrieved.
   */
  @Returns("manifestURL")
  suspend fun getManifestForFrame(@ParamName("frameId") frameId: String): String

  @EventName("applicationCacheStatusUpdated")
  fun onApplicationCacheStatusUpdated(eventListener: EventHandler<ApplicationCacheStatusUpdated>): EventListener

  @EventName("applicationCacheStatusUpdated")
  fun onApplicationCacheStatusUpdated(eventListener: suspend (ApplicationCacheStatusUpdated) -> Unit): EventListener

  @EventName("networkStateUpdated")
  fun onNetworkStateUpdated(eventListener: EventHandler<NetworkStateUpdated>): EventListener

  @EventName("networkStateUpdated")
  fun onNetworkStateUpdated(eventListener: suspend (NetworkStateUpdated) -> Unit): EventListener
}
